File: lit_nlp/client/default/layout.ts
/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Import Modules
import '../modules/span_graph_module';

import {LitModuleType} from '../core/lit_module';
import {LitComponentLayouts} from '../lib/types';
import {AnnotatedTextGoldModule, AnnotatedTextModule} from '../modules/annotated_text_module';
import {AttentionModule} from '../modules/attention_module';
import {ClassificationModule} from '../modules/classification_module';
import {ColorModule} from '../modules/color_module';
import {ConfusionMatrixModule} from '../modules/confusion_matrix_module';
import {CounterfactualExplainerModule} from '../modules/counterfactual_explainer_module';
import {CurvesModule} from '../modules/curves_module';
import {DataTableModule, SimpleDataTableModule} from '../modules/data_table_module';
import {DatapointEditorModule, SimpleDatapointEditorModule} from '../modules/datapoint_editor_module';
import {DocumentationModule} from '../modules/documentation_module';
import {EmbeddingsModule} from '../modules/embeddings_module';
import {FeatureAttributionModule} from '../modules/feature_attribution_module';
import {GeneratedImageModule} from '../modules/generated_image_module';
import {GeneratedTextModule} from '../modules/generated_text_module';
import {GeneratorModule} from '../modules/generator_module';
import {LanguageModelPredictionModule} from '../modules/lm_prediction_module';
import {MetricsModule} from '../modules/metrics_module';
import {MultilabelModule} from '../modules/multilabel_module';
import {PdpModule} from '../modules/pdp_module';
import {RegressionModule} from '../modules/regression_module';
import {SalienceClusteringModule} from '../modules/salience_clustering_module';
import {SalienceMapModule} from '../modules/salience_map_module';
import {ScalarModule} from '../modules/scalar_module';
import {SequenceSalienceModule} from '../modules/sequence_salience_module';
import {SliceModule} from '../modules/slice_module';
import {SpanGraphGoldModuleVertical, SpanGraphModuleVertical} from '../modules/span_graph_module';
import {TCAVModule} from '../modules/tcav_module';
import {TrainingDataAttributionModule} from '../modules/tda_module';
import {ThresholderModule} from '../modules/thresholder_module';

// clang-format off
const MODEL_PREDS_MODULES: readonly LitModuleType[] = [
  SpanGraphGoldModuleVertical,
  SpanGraphModuleVertical,
  ClassificationModule,
  MultilabelModule,
  RegressionModule,
  LanguageModelPredictionModule,
  GeneratedTextModule,
  AnnotatedTextGoldModule,
  AnnotatedTextModule,
  GeneratedImageModule,
];

const DEFAULT_MAIN_GROUP: readonly LitModuleType[] = [
  DataTableModule,
  DatapointEditorModule,
  SliceModule,
  ColorModule,
];
// clang-format on

// clang-format off
/**
 * Possible layouts for LIT (component groups and settigns.)
 */
export const LAYOUTS: LitComponentLayouts = {
  /**
   * A "simple demo server" layout.
   */
  'simple':  {
    upper: {
      "Editor": [DocumentationModule, SimpleDatapointEditorModule],
      "Examples": [SimpleDataTableModule],
    },
    lower: {
      'Predictions': [ ...MODEL_PREDS_MODULES],
      'Salience': [SalienceMapModule, SequenceSalienceModule],
      'Influence': [TrainingDataAttributionModule],
    },
    layoutSettings: {
      hideToolbar: true,
      mainHeight: 30,
      centerPage: true
    },
    description: 'A basic layout just containing a datapoint creator/editor, the predictions, and the data table. There are also some visual simplifications: the toolbar is hidden, and the modules are centered on the page rather than being full width.'
  },
  /**
   * A default layout for LIT Modules
   */
  'default':  {
    components : {
      'Main': [DocumentationModule, EmbeddingsModule, ...DEFAULT_MAIN_GROUP],
      'Predictions': [
        ...MODEL_PREDS_MODULES,
        ScalarModule,
        PdpModule,
      ],
      'Explanations': [
        ...MODEL_PREDS_MODULES,
        SalienceMapModule,
        SequenceSalienceModule,
        AttentionModule,
        FeatureAttributionModule,
      ],
      'Clustering': [SalienceClusteringModule],
      'Metrics': [
        MetricsModule,
        ConfusionMatrixModule,
        CurvesModule,
        ThresholderModule,
      ],
      'Influence': [TrainingDataAttributionModule],
      'Counterfactuals': [GeneratorModule, CounterfactualExplainerModule],
      'TCAV': [
        TCAVModule,
      ],
    },
    description: "The default LIT layout, which includes the data table and data point editor, the performance and metrics, predictions, explanations, and counterfactuals."
  },
};
// clang-format on

File: lit_nlp/api/layout.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Module names and type definitions for frontend UI layouts."""
from collections.abc import Mapping, Sequence
import enum
from typing import Optional, Union

import attr
from lit_nlp.api import dtypes


# pylint: disable=invalid-name
@enum.unique
class LitModuleName(dtypes.EnumSerializableAsValues, enum.Enum):
  """List of available frontend modules.

  Entries should map the TypeScript class name to the HTML element name,
  as declared in HTMLElementTagNameMap in the .ts file defining each LitModule.
  """
  # keep-sorted start
  AttentionModule = 'attention-module'
  ClassificationModule = 'classification-module'
  ConfusionMatrixModule = 'confusion-matrix-module'
  CurvesModule = 'curves-module'
  DataTableModule = 'data-table-module'
  DatapointEditorModule = 'datapoint-editor-module'
  DiveModule = 'dive-module'
  EmbeddingsModule = 'embeddings-module'
  FeatureAttributionModule = 'feature-attribution-module'
  GeneratedTextModule = 'generated-text-module'
  GeneratorModule = 'generator-module'
  LegacySequenceSalienceModule = 'legacy-sequence-salience-module'
  MetricsModule = 'metrics-module'
  MultilabelModule = 'multilabel-module'
  PdpModule = 'pdp-module'
  SalienceMapModule = 'salience-map-module'
  ScalarModule = 'scalar-module'
  SequenceSalienceModule = 'sequence-salience-module'
  SimpleDataTableModule = 'simple-data-table-module'
  # Simplified, non-replicating version of Datapoint Editor
  SimpleDatapointEditorModule = 'simple-datapoint-editor-module'
  # Non-replicating version of Datapoint Editor
  SingleDatapointEditorModule = 'single-datapoint-editor-module'
  ThresholderModule = 'thresholder-module'
  TrainingDataAttributionModule = 'tda-module'
  # keep-sorted end

  def __call__(self, **kw):
    return ModuleConfig(self.value, **kw)


# LINT.IfChange
# TODO(lit-dev): consider making modules subclass this instead of LitModuleName.
@attr.s(auto_attribs=True)
class ModuleConfig(dtypes.DataTuple):
  module: Union[str, LitModuleName]
  requiredForTab: bool = False
  # TODO(b/172979677): support title, duplicateAsRow, numCols,
  # and startMinimized.


# Most users should use LitModuleName, but we allow fallback to strings
# so that users can reference custom modules which are defined in TypeScript
# but not included in the LitModuleName enum above.
# If a string is used, it should be the HTML element name, like foo-bar-module.
LitModuleList = Sequence[Union[str, LitModuleName, ModuleConfig]]
# Keys are names of tabs, and values are names of LitModule HTML elements, e.g.,
# data-table-module for the DataTableModule class.
LitTabGroupLayout = Mapping[str, LitModuleList]


@attr.s(auto_attribs=True)
class LayoutSettings(dtypes.DataTuple):
  hideToolbar: bool = False
  mainHeight: int = 45
  leftWidth: int = 50
  centerPage: bool = False


@attr.s(auto_attribs=True)
class LitCanonicalLayout(dtypes.DataTuple):
  """Frontend UI layout; should match client/lib/types.ts."""
  upper: LitTabGroupLayout
  lower: LitTabGroupLayout = attr.ib(factory=dict)
  left: LitTabGroupLayout = attr.ib(factory=dict)
  layoutSettings: LayoutSettings = attr.ib(factory=LayoutSettings)
  description: Optional[str] = None

  def to_json(self) -> dtypes.JsonDict:
    """Override serialization to properly convert nested objects."""
    # Not invertible, but these only go from server -> frontend anyway.
    return attr.asdict(self, recurse=True)


LitComponentLayouts = Mapping[str, LitCanonicalLayout]

# pylint: enable=invalid-name
# LINT.ThenChange(../client/lib/types.ts)

##
# Common layout definitions.

modules = LitModuleName  # pylint: disable=invalid-name

MODEL_PREDS_MODULES = (
    modules.ClassificationModule,
    modules.MultilabelModule,
    modules.GeneratedTextModule,
)

DEFAULT_MAIN_GROUP = (
    modules.DataTableModule,
    modules.DatapointEditorModule,
)

##
# A "simple demo server" layout.
SIMPLE_LAYOUT = LitCanonicalLayout(
    upper={
        'Editor': [
            modules.SimpleDatapointEditorModule,
        ],
        'Examples': [modules.SimpleDataTableModule],
    },
    lower={
        'Predictions': list(MODEL_PREDS_MODULES),
        'Salience': [
            *MODEL_PREDS_MODULES,
            modules.SalienceMapModule(requiredForTab=True),
        ],
        'Sequence Salience': [
            *MODEL_PREDS_MODULES,
            modules.LegacySequenceSalienceModule(requiredForTab=True),
        ],
        'Influence': [modules.TrainingDataAttributionModule],
    },
    layoutSettings=LayoutSettings(
        hideToolbar=True,
        mainHeight=30,
        centerPage=True,
    ),
    description=(
        'A basic layout just containing a datapoint creator/editor, the '
        'predictions, and the data table. There are also some visual '
        'simplifications: the toolbar is hidden, and the modules are centered '
        'on the page rather than being full width.'
    ),
)

THREE_PANEL_LAYOUT = LitCanonicalLayout(
    left={
        'Tabular Exploration': [modules.DataTableModule],
        'Current Example': [modules.DatapointEditorModule],
        'Visual Exploration': [modules.DiveModule],
        'Embeddings': [modules.EmbeddingsModule],
    },
    upper={
        'Predictions': MODEL_PREDS_MODULES,
        'Current Example': [modules.DatapointEditorModule],
        'Counterfactuals': [modules.GeneratorModule],
    },
    lower={
        'Metrics': [
            modules.MetricsModule,
            modules.ConfusionMatrixModule,
            modules.ThresholderModule,
        ],
        'Charts': [
            modules.ScalarModule,
            modules.PdpModule,
            modules.CurvesModule,
        ],
        'Explanations': [
            modules.SalienceMapModule,
            modules.LegacySequenceSalienceModule,
            modules.AttentionModule,
            modules.FeatureAttributionModule,
        ],
        'Influence': [modules.TrainingDataAttributionModule],
    },
    description=(
        'A three-panel layout with tools for exploring data in the aggregate or'
        ' per-example (on the left) or reviewing prediction results (upper'
        ' right) and performance characteristics, etc. (lower left).'
    ),
)

##
# A "kitchen sink" layout with maximum functionality.
STANDARD_LAYOUT = LitCanonicalLayout(
    upper={
        'Main': [
            modules.EmbeddingsModule,
            *DEFAULT_MAIN_GROUP,
        ]
    },
    lower={
        'Predictions': [
            *MODEL_PREDS_MODULES,
            modules.ScalarModule,
            modules.PdpModule,
        ],
        'Explanations': [
            *MODEL_PREDS_MODULES,
            modules.SalienceMapModule,
            modules.LegacySequenceSalienceModule,
            modules.AttentionModule,
            modules.FeatureAttributionModule,
        ],
        'Metrics': [
            modules.MetricsModule,
            modules.ConfusionMatrixModule,
            modules.CurvesModule,
            modules.ThresholderModule,
        ],
        'Influence': [modules.TrainingDataAttributionModule],
        'Counterfactuals': [
            modules.GeneratorModule,
        ],
    },
    description=(
        'The default LIT layout, which includes the data table and data point '
        'editor, the performance and metrics, predictions, explanations, and '
        'counterfactuals.'
    ),
)

DEFAULT_LAYOUTS = {
    'simple': SIMPLE_LAYOUT,
    'default': STANDARD_LAYOUT,
    'three_panel': THREE_PANEL_LAYOUT,
}
